### PR TITLE
Pr/1.4.5

### PR DESCRIPTION
**Pull request by mpvanderschelling**
This pull request will update the `main` code to version 1.4.5

### `Domain` and `Parameter` classes
* All `Parameter` classes are now private. The suggested way to add parameters to a Domain will be with the public `add_float`, `add_int`, `add_category` and `add_constant` methods.
* Change the behaviour of the public `Domain.add` method to require the type and arguments of the (private) parameter to construct, instead of requiring the private Parameter object

### `ExperimentData` class
* Added a method that extracts all `ExperimentSample` objects or a selection of `ExperimentSample` objects from an `ExperimentData` object
* Removed the `.filename` attribute in `ExperimentData`: the name of the stored objects are hardcoded
* Refactored the `.path` attribute to the `.project_dir` attribute. When storing the `ExperimentData` object with the `.store()` method, a subdirectory (`/experiment_data/`) will be created and the input data, output data, domain and jobs will be stored there.

### `Optimizer` class
* Added functionality to select different strategies to determine the initial points ($x_0$) before optimization.

### `_Data` class
* Decoupled the headers of the `_Data` class to a separate private `_Columns` class. This makes it more flexible for future development on the generalization of different back-ends (currently pandas) for `_Data`
* Removed dependency of `Domain` to `_Data` class

### Storing output data
* Output that is stored to disk will no longer have the `path_` prefix before the parameter. This information is stored within the `_Columns` object
* The `ExperimentSample.output_data` will, if applicable, load the output parameters from disk automatically
* The attribute `ExperimentSample.output_data_with_references` will give you the paths of the references
* Added backwards compatibility for `ExperimentSample.output_data_loaded` (will redirect to `ExperimentSample.output_data`

### Documentation
* Made private methods of public objects visible in sphinx documentation API 

### Bessa Group Coding Guidelines
* Restored the original .flake8 file of the BRG coding guidelines; the maximum linewidth is now 79 instead of 119.

### Miscellaneous
* Represent all private attributes and methods with an underscore (`_`).

## Closed issues

* #224 
* #206
* #225 
* #227 
* #61 
* #62 
* #73 
* #129 
* #202

### CODE REVIEW

File: docs/source/rst_doc_files/classes/design/experimentsample.rst [hunk 1]
```diff
@@ -107,29 +107,26 @@ A reference (:code:`Path`) will be saved to the :attr:`~f3dasm.design.Experiment
...
+       └── jobs.pkl
 
-In the :attr:`~f3dasm.design.ExperimentData.output_data`, a reference to the stored object (e.g. :code:`my_project/output_1/0.npy`) will be automatically appended to the `path_<output parameter name>` parameter.
-
-.. code-block:: python
-
-    >>> experiment_sample['output_numpy']
-    'my_project/output_numpy/0.npy'
 
+In the :attr:`~f3dasm.design.ExperimentData.output_data`, a reference to the stored object (e.g. :code:`/output_numpy/0.npy`) will be automatically appended to the parameter.
```
Comment (BernardoFerreira) [inline docs/source/rst_doc_files/classes/design/experimentsample.rst:129]: Can you be more clear here? I don't understand what is being appended to what...

File: src/f3dasm/_src/datageneration/abaqus/abaqus_functions.py [hunk 2]
```diff
@@ -3,9 +3,11 @@
 
 Note
 ----
-I'm assuming that every process could have a different pre-processing, execution and post-processing function.
+I'm assuming that every process could have a different pre-processing,
+ execution and post-processing function.
```
Comment (BernardoFerreira) [inline src/f3dasm/_src/datageneration/abaqus/abaqus_functions.py:7]: This leading whitespace is not required to comply with flake8. Fix all occurrences (not required for this PR).

File: src/f3dasm/_src/datageneration/functions/adapters/augmentor.py [hunk 3]
```diff
@@ -105,36 +105,47 @@ def reverse_augment(self, output: np.ndarray) -> np.ndarray:
...
 
 
 class FunctionAugmentor:
-    """Combination of Augmentors that can change the input and output of an objective function
+    """Combination of Augmentors that can change the input and
+     output of an objective function
 
     Args:
         input_augmentors (List[Augmentor]): list of input augmentors
         output_augmentors (List[Augmentor]): list of output augmentors
```
Comment (BernardoFerreira) [inline src/f3dasm/_src/datageneration/functions/adapters/augmentor.py:141]: Wrong docstring formatting type.

File: src/f3dasm/_src/experimentdata/_io.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+INPUT_DATA_FILENAME = "input"
+OUTPUT_DATA_FILENAME = "output"
+JOBS_FILENAME = "jobs"
+
+#                                                               Storing methods
+# =============================================================================
+
+class _Store:
+    suffix: int
+
```
Comment (BernardoFerreira) [inline src/f3dasm/_src/experimentdata/_io.py:47]: There are a lot of docstrings missing in this new module. Please add them, even if the methods are simple and/or private.

**Review by BernardoFerreira — CHANGES_REQUESTED** [review]
Good job with this new release version, a lot of major updates that will significantly improve f3dasm functionality for sure! Please review the few comments that I left before approval.

Thank you for the less "functionality" improvements such as the compliance with the BRG guidelines and the private leading _ convention. Nonetheless, I point out that there are some modules that still don't have the correct BRG header format and several functions/methods/classes that don't have docstrings (namely private ones). These should be fixed for the next CRW (December 5).

File: src/f3dasm/_src/datageneration/abaqus/abaqus_functions.py (hunk 2 above)
Comment (mpvanderschelling) [reply src/f3dasm/_src/datageneration/abaqus/abaqus_functions.py:7]: I've added docstrings for all functions and classes in this module!

File: src/f3dasm/_src/datageneration/functions/adapters/augmentor.py (hunk 3 above)
Comment (mpvanderschelling) [reply src/f3dasm/_src/datageneration/functions/adapters/augmentor.py:141]: Fixed it!

**Review by BernardoFerreira — APPROVED** [review]

**Pull request merged**